From 350434eccc0bcfe49d3e63046131df44970b1b58 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <michael.i.doherty@intel.com>
Date: Thu, 8 Sep 2016 14:32:54 +0100
Subject: [PATCH 4/4] common: Support a stateless configuration for
 /etc/lightdm/users.conf

With this change we can support both the admin-defined users.conf, as well
as a vendor-provided users.conf in /usr/share/lightdm/. This will permit
vendor configurations to provide the default sane settings, in instances
where it is not desirable to provide AccountsService in a lighter
deployment.

In the instance that the admin defined file exists, it will be used instead
of the vendor-config.

Signed-off-by: Ikey Doherty <michael.i.doherty@intel.com>
---
 common/user-list.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/common/user-list.c b/common/user-list.c
index 3169853..b213dfc 100644
--- a/common/user-list.c
+++ b/common/user-list.c
@@ -166,6 +166,7 @@ G_DEFINE_TYPE (CommonSession, common_session, G_TYPE_OBJECT);
 
 #define PASSWD_FILE      "/etc/passwd"
 #define USER_CONFIG_FILE "/etc/lightdm/users.conf"
+#define VENDOR_USER_CONFIG_FILE "/usr/share/lightdm/users.conf"
 
 static CommonUserList *singleton = NULL;
 
@@ -333,13 +334,20 @@ load_passwd_file (CommonUserList *user_list, gboolean emit_add_signal)
     gchar **hidden_users, **hidden_shells;
     GList *users = NULL, *old_users, *new_users = NULL, *changed_users = NULL, *link;
     GError *error = NULL;
+    const gchar *user_config_file = NULL;
 
-    g_debug ("Loading user config from %s", USER_CONFIG_FILE);
+    /* Load /etc/ first, otherwise fallback to /usr/share for stateless operation */
+    if (g_file_test(USER_CONFIG_FILE, G_FILE_TEST_EXISTS))
+        user_config_file = USER_CONFIG_FILE;
+    else
+        user_config_file = VENDOR_USER_CONFIG_FILE;
+
+    g_debug ("Loading user config from %s", user_config_file);
 
     config = g_key_file_new ();
-    g_key_file_load_from_file (config, USER_CONFIG_FILE, G_KEY_FILE_NONE, &error);
+    g_key_file_load_from_file (config, user_config_file, G_KEY_FILE_NONE, &error);
     if (error && !g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
-        g_warning ("Failed to load configuration from %s: %s", USER_CONFIG_FILE, error->message);
+        g_warning ("Failed to load configuration from %s: %s", user_config_file, error->message);
     g_clear_error (&error);
 
     if (g_key_file_has_key (config, "UserList", "minimum-uid", NULL))
-- 
2.10.0

